pkg/util/k8sutil/k8sutil.go

```diff
@@ -597,3 +592,166 @@ func UniqueMemberName(clusterName string) string {
 	}
 	return clusterName + "-" + suffix
 }
+
+func AttemptRestoreFromBackup(ctx context.Context, namespace string, kubecli kubernetes.Interface, cluster *api.EtcdCluster, logger *logrus.Entry) error {
+
+	latestBackup, err := getLatestBackup(ctx, namespace, kubecli, cluster, logger)
+	if err != nil {
+		return fmt.Errorf("getting latest backup failed : %v", err)
+	}
+
+	logger.Infof("attempting restore from backup : %s", latestBackup)
+	command := []string{"restore_from_backup", getBackupProjectName(cluster.Name), latestBackup}
+	output, err := execIntoBackupPod(ctx, namespace, kubecli, cluster, "util", command, logger)
+	if err != nil {
+		return fmt.Errorf("failed execing into backup pod: %v", err)
+	}
+
+	scanner := bufio.NewScanner(strings.NewReader(output))
+	for scanner.Scan() {
+		logger.Infof("stdout from restore: %s", scanner.Text())
+	}
+
+	return nil
+}
+
+func getBackupPodName(ctx context.Context, kubecli kubernetes.Interface) (string, error) {
+
+	options := metav1.ListOptions{
+		LabelSelector: "app.kubernetes.io/name=cray-etcd-backup",
+	}
+	podList, err := kubecli.CoreV1().Pods("operators").List(ctx, options)
+
+	if err != nil {
+		return "", err
+	}
+	for i := range podList.Items {
+		pod := &podList.Items[i]
+		if pod.Status.Phase == v1.PodRunning {
+			return pod.Name, nil
+		}
+	}
+	return "", fmt.Errorf("failed to get backup pod")
+}
+
+func getLatestBackup(ctx context.Context, namespace string, kubecli kubernetes.Interface, cluster *api.EtcdCluster, logger *logrus.Entry) (string, error) {
+
+	command := []string{"list_backups", getBackupProjectName(cluster.Name)}
+	output, err := execIntoBackupPod(ctx, namespace, kubecli, cluster, "boto3", command, logger)
+	if err != nil {
+		return "", fmt.Errorf("failed execing into backup pod: %v", err)
+	}
+
+	const timeLayout = "2006-01-02-15:04:05"
+	var latestBackup string
+	scanner := bufio.NewScanner(strings.NewReader(output))
+	for scanner.Scan() {
+		split := strings.Split(scanner.Text(), "/")
+		thisBackup := split[1]
+		thisTsStr := parseBackupTimeStamp(thisBackup)
+		if latestBackup == "" {
+			latestBackup = thisBackup
+		} else {
+			thisTs, _ := time.Parse(timeLayout, thisTsStr)
+			latestTsStr := parseBackupTimeStamp(latestBackup)
+			latestTs, _ := time.Parse(timeLayout, latestTsStr)
+			if thisTs.After(latestTs) {
+				latestBackup = thisBackup
+			}
+		}
+	}
+
+	return latestBackup, nil
+}
```

Comment:
As I read this function, it appears that it is selecting the "latest" backup based on the timestamp embedded in the backup name.  This means that if the admin manually creates a backup that doesn't match the expected naming conventions, it will not be considered in this routine.  I would much rather prefer that this code fetch the modified times from the S3 object, and use the latest backup based on that time, and not the one embedded in the name.  It's not clear to me what the `parseBackupTimeStamp()` routine will do if the backup name doesn't match the expected format.

In Mercury, the automatic restore from backup for failed clusters lists the available backups and assumes they are listed in date order (which has been our S3 observation) and takes the last one in the list.

Reply:
great point -- i wasn't relying on the order returned since i've seen the version part of the backup can make that unreliable.  i like the idea of ensuring we're grabbing the absolute latest based on S3 modified time.  i'll do some testing with manual backups..

Reply:
@dborman-hpe great catch, the code (before last commit) wouldn't properly handle manual backups.  First, below is an example where the default listing of the backups won't always result in the latest being the last.  The version portion of the string doesn't always increase, in addition to manual backups not having a time stamp:

```
cray-bss/brad-manual-backup
cray-bss/brad-manual-backup-1
cray-bss/brad-manual-backup-2
cray-bss/brad-manual-backup-3
cray-bss/brad-manual-backup-5
cray-bss/etcd.backup_v2285_2022-08-18-21:12:10
cray-bss/etcd.backup_v2285_2022-08-18-21:13:09
cray-bss/etcd.backup_v2288_2022-08-18-21:14:09
cray-bss/etcd.backup_v2291_2022-08-18-21:15:09
cray-bss/etcd.backup_v2294_2022-08-18-21:16:09
cray-bss/etcd.backup_v2297_2022-08-18-21:17:09
cray-bss/etcd.backup_v2303_2022-08-18-20:59:01
```

Here's a test (with the new code) where the latest is brad-manual-backup-6:

```
ncn-m001-22703c05:/home/bklein # kubectl get etcdbackups.etcd.database.coreos.com -n services cray-bss-etcd-cluster-manual-backup-28870 -o json | jq -r '(.spec.s3.path, .status.lastSuccessDate)'
etcd-backup/cray-bss/brad-manual-backup-6
2022-08-18T21:21:56Z
```

```
cray-bss/brad-manual-backup
cray-bss/brad-manual-backup-1
cray-bss/brad-manual-backup-2
cray-bss/brad-manual-backup-3
cray-bss/brad-manual-backup-5
cray-bss/brad-manual-backup-6
cray-bss/etcd.backup_v2285_2022-08-18-21:13:09
cray-bss/etcd.backup_v2288_2022-08-18-21:14:09
cray-bss/etcd.backup_v2291_2022-08-18-21:15:09
cray-bss/etcd.backup_v2294_2022-08-18-21:16:09
cray-bss/etcd.backup_v2297_2022-08-18-21:17:09
cray-bss/etcd.backup_v2299_2022-08-18-21:20:02
cray-bss/etcd.backup_v2302_2022-08-18-21:21:02
```

And it gets selected:

```
time="2022-08-18T21:22:37Z" level=info msg="attempting restore from backup : brad-manual-backup-6" cluster-name=cray-bss-etcd cluster-namespace=services pkg=cluster
```